### PR TITLE
GitHub Workflow Documentation

### LINKED ISSUE #23: Create docs page Reference/GitHub/Workflow
### Purpose
Document GitHub workflow

### Page Tree
```yml
nav:
  ...
  - Reference:
    ...
    - GitHub:
      - Workflow: reference/github/workflow.md
```

### Contents
- Creating issue
    - Using issue templates (or creating from scratch)
    - Adding to project
    - Adding labels (label conventions)
    - Adding assignees
- Creating branch, pushing commits
    - Include Git commands
- Creating PR
    - Linking issue or adding to project
    - Adding labels (label conventions)
    - Adding reviewers
    - Create as draft until ready to be reviewed
- Merging into master
    - Requirements
        - Pass CI
        - Approved by reviewers

### Relevant Resources
- https://patrick-5546.github.io/notes/reference/git/git_contributing/
- https://patrick-5546.github.io/notes/reference/git/git_commands/

### PR DESCRIPTION
**Pull request by DFriend01**
Resolves #23.

Please review the pages under Reference/GitHub/Workflow. In particular, check for the following:

1. The content is clear and concise
2. There is just the right amount of information and we are not missing anything
3. Spelling and grammar

Refer to the issue description for the expected content. I did add a little more than what was on the issue description, including:

- Git merge and resolving merge conflicts
- A workflow summary
- Git learning resources

### CODE REVIEW

File: docs/reference/github/workflow/index.md
Comment (patrick-5546) [inline docs/reference/github/workflow/index.md:15]: I think a good development workflow increases team efficiency and ensures code quality

File: docs/reference/github/workflow/index.md
Comment (patrick-5546) [inline docs/reference/github/workflow/index.md]: - Don't need to prefix with "GitHub"
- Link to our doc page if there is one for that feature
- Can explicitly list the features that have their own tab: issues, pull requests, actions
    - We could talk about other features more in #24

File: docs/reference/github/workflow/index.md
Comment (patrick-5546) [inline docs/reference/github/workflow/index.md:13]: This is beautiful 😄

File: docs/reference/github/workflow/index.md
Comment (patrick-5546) [inline docs/reference/github/workflow/index.md]: Could rename to `Version control: Git`

File: docs/reference/github/workflow/index.md
Comment (patrick-5546) [inline docs/reference/github/workflow/index.md]: Could rename to `Remote server: GitHub`

File: docs/reference/github/workflow/issues.md
Comment (patrick-5546) [inline docs/reference/github/workflow/issues.md]: Maybe delete the second sentence, it's pretty technical for the overview

File: docs/reference/github/workflow/issues.md
Comment (patrick-5546) [inline docs/reference/github/workflow/issues.md:23]: You could add a note to link to where you can add or edit the issue templates: https://github.com/UBCSailbot/.github/issues/templates/edit

File: docs/reference/github/workflow/issues.md
Comment (patrick-5546) [inline docs/reference/github/workflow/issues.md]: Maybe we shouldn't document this here as the template sections could change. One thing that we can do is add section descriptions in the templates as comments. We can discuss this at our next infrastructure meeting

File: docs/reference/github/workflow/issues.md
Comment (patrick-5546) [inline docs/reference/github/workflow/issues.md:85]: Can add a note about how issue templates will often automatically add the relevant labels

File: docs/reference/github/workflow/branches.md
Comment (patrick-5546) [inline docs/reference/github/workflow/branches.md]: To simplify things we could say to only create branches from the main branch (can discuss this in next infra meeting)

```
git switch main
git switch -c <branch_name>
```

Notes: `switch` and `restore` was introduced in newer version of git to replace `checkout`: https://stackoverflow.com/a/57266005

File: docs/reference/github/workflow/branches.md
Comment (patrick-5546) [inline docs/reference/github/workflow/branches.md:30]: Is there any benefit in having a branch naming convention? As far as I know there is no way to enforce one on GitHub and its something extra to remember. Can discuss this next infra meeting.

File: docs/reference/github/workflow/branches.md
Comment (patrick-5546) [inline docs/reference/github/workflow/branches.md:73]: Should we also talk about rebasing and when to use merging vs rebasing, or is that outside the scope of this page?

File: docs/reference/github/workflow/branches.md
Comment (patrick-5546) [inline docs/reference/github/workflow/branches.md:110]: - May be better to put in pull request page
- Can also resolve merge conflicts on GitHub (recommended) or VS Code

File: docs/reference/github/workflow/pr.md
Comment (patrick-5546) [inline docs/reference/github/workflow/pr.md]: I would say that PR's are used to verify code functionality and quality of a development branch before merging into the main branch, accomplished through CI and code reviews.

File: docs/reference/github/workflow/pr.md
Comment (patrick-5546) [inline docs/reference/github/workflow/pr.md:22]: Should the main branch always be the base branch? Relates to another comment above that we should discuss

File: docs/reference/github/workflow/pr.md
Comment (patrick-5546) [inline docs/reference/github/workflow/pr.md]: I think you need a closing keyword as well: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

File: docs/reference/github/workflow/branches.md
Comment (patrick-5546) [inline docs/reference/github/workflow/branches.md:1]: Can add a pro tips section
- recommended git config (I have some ideas on this)
- git aliases ideas: https://github.com/ohmyzsh/ohmyzsh/blob/master/plugins/git/README.md#aliases

File: docs/reference/github/workflow/pr.md
Comment (patrick-5546) [inline docs/reference/github/workflow/pr.md:67]: Would it make more sense to have this section last? It is from the perspective of the reviewer rather than creator like the other sections

File: docs/reference/github/workflow/pr.md
Comment (patrick-5546) [inline docs/reference/github/workflow/pr.md]: Usually you wouldn't review your own PR. You could add documentation in the PR description or subsequent comments though

File: docs/reference/github/workflow/pr.md
Comment (patrick-5546) [inline docs/reference/github/workflow/pr.md]: Can expand on this, maybe in a new section (GitHub Actions)

File: docs/reference/github/workflow/pr.md
Comment (patrick-5546) [inline docs/reference/github/workflow/pr.md:12]: Can also talk about creating pull requests as drafts and when to do so: when it is still in active development (aka you don't want code reviews) but you want to check if CI passes

File: docs/reference/github/workflow/pr.md
Comment (patrick-5546) [reply docs/reference/github/workflow/pr.md]: Different types of actions
- tests: lint, unit, regression
- deploy
- there's probably more that I can't think of right now

File: docs/reference/github/workflow/pr.md
Comment (patrick-5546) [inline docs/reference/github/workflow/pr.md]: You could add stuff about adding comments as part of a review, approve vs comment vs request changes

**Review by patrick-5546 — CHANGES_REQUESTED** [review]
Amazing work, I love your documentation style.

I left a bunch of comments and did some proofreading (see my commit below. If you have any questions reply to the comment or we can discuss it next infra meeting

One thing that came to mind: should we add a References section at the bottom of each page for resources that you used? They could be helpful for further reading

File: docs/reference/github/workflow/pr.md
Comment (patrick-5546) [inline docs/reference/github/workflow/pr.md]: One more thing, you could link to where you talk about a field in the issues sectio; for example `[Labels](./issues.md#labelling-issues)`

File: docs/reference/github/workflow/index.md
Comment (DFriend01) [reply docs/reference/github/workflow/index.md:13]: Thank you <3 . I like it as well!

File: docs/reference/github/workflow/index.md
Comment (DFriend01) [reply docs/reference/github/workflow/index.md]: Can you expand on that second point?

File: docs/reference/github/workflow/issues.md
Comment (DFriend01) [reply docs/reference/github/workflow/issues.md]: Good point. I will reword the paragraph here to tell the reader to fill out the sections that they see.

File: docs/reference/github/workflow/issues.md
Comment (DFriend01) [reply docs/reference/github/workflow/issues.md:85]: You mean labels and not issues, right?

File: docs/reference/github/workflow/branches.md
Comment (DFriend01) [reply docs/reference/github/workflow/branches.md:1]: I think this should be more of a Git-specific thing rather than putting it on the branches page. It seems a little out of place. We could consider adding our own Git reference in the future and add these sorts of things to it.

File: docs/reference/github/workflow/branches.md
Comment (DFriend01) [reply docs/reference/github/workflow/branches.md]: I think that we should still keep this information here because this allows developers to create their own local branches where they can branch off ideas. I agree that we should only branch from `main` for branches **related to issues**. I think that beyond this, we should have the freedom to do whatever we want in our local workspace as long as it does not interfere with the remote repository.

I'll add a note that when creating branches for issues, we should do it from `main`, but I still think that we should keep the information.

File: docs/reference/github/workflow/branches.md
Comment (DFriend01) [reply docs/reference/github/workflow/branches.md:30]: I think that we should enforce this because this allows us to know who is working on what issue in a specific branch. There have been far too many instances where I have seen mysterious branches and I don't even know what they were used for. It is one extra thing to remember, but I believe that it's good practice and will serve us well.

File: docs/reference/github/workflow/branches.md
Comment (DFriend01) [reply docs/reference/github/workflow/branches.md:73]: To be honest, rebasing is not a topic that I fully understand myself and is a far more difficult thing to do. I don't really see the benefit from using rebasing over merging which is why I did not include it. It's more of an advanced topic.

File: docs/reference/github/workflow/branches.md
Comment (DFriend01) [reply docs/reference/github/workflow/branches.md:110]: Merging branches is part of a pull request, but not exclusive to them. That is why I thought that the topic on merging itself should belong in this section.

File: docs/reference/github/workflow/pr.md
Comment (DFriend01) [reply docs/reference/github/workflow/pr.md:22]: This depends on if we use an additional branch for development. For example, issue --> development branch --> main branch. I think that adding another branch might make it more complicated and it's not super necessary for us since the project isn't too big with a relatively small amount of people working on it.

File: docs/reference/github/workflow/pr.md
Comment (DFriend01) [reply docs/reference/github/workflow/pr.md]: This was more of me mentioning that there are CI checks. I assumed that we have an issue for a GitHub Actions page since that is a whole beast itself, so let's leave this be.

File: docs/reference/github/workflow/pr.md
Comment (DFriend01) [reply docs/reference/github/workflow/pr.md]: My intentions for this section was to describe what you would do in a PR at a high level, not go into specifics. I did think of adding what you mentioned, but this was something that one would gain from experience with PRs (we are over documenting at this point).

File: docs/reference/github/workflow/index.md
Comment (patrick-5546) [reply docs/reference/github/workflow/index.md]: For example, `[Milestones](./issues.md#adding-issues-to-a-milestone)`

File: docs/reference/github/workflow/issues.md
Comment (patrick-5546) [reply docs/reference/github/workflow/issues.md]: - I created an issue for it: https://github.com/UBCSailbot/.github/issues/4

File: docs/reference/github/workflow/branches.md
Comment (patrick-5546) [reply docs/reference/github/workflow/branches.md]: If we are only documenting what you need for sailbot to get started I don't think the other options are necessary, but as long as the sailbot convention is front and center I'm good

File: docs/reference/github/workflow/branches.md
Comment (patrick-5546) [reply docs/reference/github/workflow/branches.md:30]: I agree, but might I suggest using one's GitHub username instead of first name.

Also, some branches might not be tied to issues so I don't think there has to be any requirements for after the `/`.
What I like to do is keep the branch name short and have a better description in the PR title

File: docs/reference/github/workflow/branches.md
Comment (patrick-5546) [reply docs/reference/github/workflow/branches.md:90]: Let's talk about this tonight

File: docs/reference/github/workflow/pr.md
Comment (patrick-5546) [reply docs/reference/github/workflow/pr.md:22]: I agree, but asked because of the wording "For the most part". I'll leave it up to you whether you want to change that.

File: docs/reference/github/workflow/pr.md
Comment (patrick-5546) [reply docs/reference/github/workflow/pr.md]: - Issue to create PR template: https://github.com/UBCSailbot/.github/issues/5

File: docs/reference/github/workflow/branches.md
Comment (patrick-5546) [reply docs/reference/github/workflow/branches.md:1]: Yea on second thought I think it is outside the scope of this site

File: docs/reference/github/workflow/pr.md
Comment (patrick-5546) [reply docs/reference/github/workflow/pr.md]: There is not yet an issue for it. We could add it to this PR or create a new issue. If we focus on what someone needs to know to pass CI I don't think it has to be that long (essentially check the logs and search for "Error")

File: docs/reference/github/workflow/pr.md
Comment (patrick-5546) [reply docs/reference/github/workflow/pr.md]: Do we even need a section for reviewing PR's then, mostly leads would be the ones reviewing, and it's pretty self explanatory

File: docs/reference/github/workflow/pr.md
Comment (DFriend01) [reply docs/reference/github/workflow/pr.md:12]: I will look into creating drafts then add it to the documentation.

File: docs/reference/github/workflow/pr.md
Comment (DFriend01) [reply docs/reference/github/workflow/pr.md]: Will do this.

File: docs/reference/github/workflow/pr.md
Comment (DFriend01) [reply docs/reference/github/workflow/pr.md:67]: Makes sense. Will put this section last and put a comment in the previous section about reviewers adding comments to their PR.

File: docs/reference/github/workflow/pr.md
Comment (DFriend01) [reply docs/reference/github/workflow/pr.md]: Good point, but it could serve new members well since they have less experience and reviewing PRs should not be exclusive to leads, but should be inclusive for everyone involved.

File: docs/reference/github/workflow/pr.md
Comment (DFriend01) [reply docs/reference/github/workflow/pr.md]: Add something short on checking if CI checks pass or fail.

File: docs/reference/github/workflow/branches.md
Comment (DFriend01) [reply docs/reference/github/workflow/branches.md]: Good point, I should probably simplify this part. I will address this.

File: docs/reference/github/workflow/branches.md
Comment (DFriend01) [reply docs/reference/github/workflow/branches.md:30]: Ok, I will change it to GitHub username and make a note on the issue number being optional.

File: docs/reference/github/workflow/branches.md
Comment (DFriend01) [reply docs/reference/github/workflow/branches.md:110]: Could consider removing this section all together since it is a complex topic. Could just link to other resource that explain it better than us.

File: docs/reference/github/workflow/branches.md
Comment (patrick-5546) [reply docs/reference/github/workflow/branches.md:73]: If merging new commits in main to development branch, might be better to rebase if you don’t want it to show up in the branch’s commit history (e.g. when it's unrelated to PR purpose). However it messes up the commits (force pushes), so I don’t oppose leaving it out of the docs

File: docs/reference/github/workflow/issues.md
Comment (DFriend01) [reply docs/reference/github/workflow/issues.md:23]: I had to remove this because the link checker considers this a dead link.

File: docs/reference/github/workflow/issues.md
Comment (DFriend01) [reply docs/reference/github/workflow/issues.md]: I copied that sentence from the git documentation. They don't seem to think it's an issue.

File: docs/reference/github/workflow/issues.md
Comment (patrick-5546) [inline docs/reference/github/workflow/issues.md]: Instead of removing this link because it doesn't pass the link check, we can ignore it in `.markdown-link-check.json`

Comment (patrick-5546) [reply docs/reference/github/workflow/issues.md]: For example

```json
  "ignorePatterns": [
    {
      "pattern": "^http://example.net"
    }
  ],
```

**Review by patrick-5546 — APPROVED** [review]

**Pull request merged**
